### DISCUSSION TITLE
Docker Compose Errors

### DISCUSSION THREAD

**Discussion by FlaMike** [tagyoureit/nodejs-poolController#674] [Q&A]
I am trying to migrate the njsPC and the Dashboard from a Pi to a Docker system running under Ubuntu Server 22.04.  I verified the port for the adapter.  I get the following error messages when starting njsPC

    at Module.require (node:internal/modules/cjs/loader:1005:19)
Error reading configuration information.  Aborting startup: Error: EISDIR: illegal operation on a directory, read
Error: EISDIR: illegal operation on a directory, read
    at Object.readSync (node:fs:723:3)
    at tryReadSync (node:fs:433:20)
    at Object.readFileSync (node:fs:479:19)
    at new Config (/app/config/Config.ts:37:69)
    at Object.<anonymous> (/app/config/Config.ts:192:31)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)

And I get the following error messages when trying to start the Dashboard:

Error reading configuration information.  Aborting startup: Error: EISDIR: illegal operation on a directory, read
/app/dist/server/config/Config.js:27
            throw err;
            ^
Error: EISDIR: illegal operation on a directory, read
    at Object.readSync (node:fs:723:3)
    at tryReadSync (node:fs:433:20)
    at Object.readFileSync (node:fs:479:19)
    at new Config (/app/dist/server/config/Config.js:16:69)
    at Object.<anonymous> (/app/dist/server/config/Config.js:106:18)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19) {
  errno: -21,
  syscall: 'read',
  code: 'EISDIR'
}

Do you have any suggestions as  to how to repair the errors?  Thanks!

**Comment by tagyoureit** [comment]
Have you gone through the [wiki](https://github.com/tagyoureit/nodejs-poolController/wiki/Docker)?  You have to map the writeable directories to a physical drive.

**Reply by FlaMike** [reply]
I did update the volumes section in the docker compose file.

**Reply by tagyoureit** [reply]
Can to post up your docket-compose file?  You may have a '/' inadvertently after a file name which could be interpreted as a directory.

**Reply by FlaMike** [reply]
I hope it is something simple as you suggest.  Here it is.  Thank you for your help!

  
![image](https://user-images.githubusercontent.com/41341956/210132434-fe1dfbd0-9d3b-4cc4-9b9d-2debb19b8479.png)

**Reply by FlaMike** [reply]
I was able to get the pool controller and the dashboard working.  Here is what I did...

I created the poolcontroller and its associated data directories.  I then did a "touch config.jason" in the poolcontroller directory.  This created an empty config.jason file.

I did the same for the -dashboard directory and its config.json file.

I copied the config.jason files from the Pi where the programs were running.

I edited the dashpanels config.json to point to the new server IP address.

I did a chmod 777 on both config.json files, their parent directories, and the data directories.

I rebuilt the containers (docker compose up -d)

I checked the logs for both programs & all looked good (other than the latitude/longitude error).  I was then able to open the dashboard & update the latitude and longitude.  The resulting time errors went away.

The programs found the EasyTouch, populated the equipment & I am able to operate the features from the dashpanel.

I'm using Crewski's HA integration.  I am getting data for the pool equipment & sensors but at this time, I am not able to control them through HA.  I will drop him a note & see if he has any suggestions.

Not quite there yet, but overall, great progress! 

I wanted to provide a reasonably detailed explanation, because none of what was needed was obvious or in the wiki.  I did not make any changes to the docker compose file from the one posted above.  It's nearly identical to the wiki, other then the device specification and the directory roots.

**Reply by rstrouse** [reply]
Hmmm so config.json is dynamically created.  We got a chicken/egg thing going on where the file doesn't yet exist to set the permissions but it is needed for the configuration.  If we add it to the repo it will get overwritten during the pull.

**Reply by FlaMike** [reply]
I don't know a way around it (still very new to docker) but a sidebar in the docs may help others in the future.  More knowledgeable folks than me may be able to provide a more elegant solution.

BTW, I solved the problem with crewski's HA integration.  I wrote it up in his github, but in short, it was pilot error on my part.  I forgot to turn off the Pi when I moved the 485 dongle to the docker box.  Apparently, the Pi could get some (stale?) data from the pool controller but since it had no 485 connection it could not issue commands & get status updates.  Turning off the Pi, reloading the integrations, restarting HA, & providing the integration with the (new) IP address of the controller did the trick.

What a day.  I'm beat but happy :-)
